File: ios/Infra/Adapters/Rar/RarExtractor.swift
//
//  RarExtractor.swift
//  uncompress
//
//  Created by Adriano Souza Costa on 15/03/21.
//

import Foundation
import UIKit
import UnrarKit

public final class RarExtractor: Extractable {
    
    public var compatibilities: [Compatibility] {
        return [.rar, .cbr]
    }
    
    public init() {}
    
    public func extract(
        _ filePath: URL,
        to destination: URL,
        overwrite: Bool = true,
        password: String? = nil
    ) throws {
        let archive = try URKArchive(path: filePath.absoluteString)
        
        if let password = password {
            archive.password = password
        }
        
        try archive.extractFiles(to: destination.absoluteString, overwrite: overwrite)
    }
    
}

File: android/src/main/java/com/uncompress/infra/adapters/rar/RarExtractor.kt
package com.uncompress.infra.adapters.rar

import com.github.junrar.Archive
import com.github.junrar.Junrar
import com.uncompress.domain.enum.Compatibility
import com.uncompress.data.intefaces.Extractable
import java.io.File

final class RarExtractor: Extractable {

  override val compatibilities: List<Compatibility>
    get() = listOf(Compatibility.CBR, Compatibility.RAR)

  @Throws(Throwable::class)
  override fun extract(
    filePath: String,
    destination: String,
    override: Boolean,
    password: String?
  ) {
    super.extract(filePath, destination, override, password)
    Junrar.extract(filePath, destination, password)
  }

  @Throws(Throwable::class)
  override fun isProtected(filePath: String): Boolean {
    val archive = Archive(File(filePath))
    return archive.isEncrypted
  }
}
